scripts/metricsgen/metricsgen.go

```diff
@@ -0,0 +1,314 @@
+// metricsgen is a code generation tool for creating constructors for Tendermint
+// metrics types.
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/parser"
+	"go/token"
+	"go/types"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strings"
+	"text/template"
+)
+
+func init() {
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, `Usage: %[1]s -dir <dir> -struct <struct>
+
+Generate constructors for the metrics type contained in the specified -dir
+Go directory. The tool creates a new file in the same directory as the specified directory
+containing the generated code.
+
+Options:
+`, filepath.Base(os.Args[0]))
+		flag.PrintDefaults()
+	}
+}
+
+const (
+	metricNameTag = "metricsgen_name"
```

Comment:
I'd suggest trimming the `sgen` off these names, e.g., `metric_labels`, `metric_buckets`, etc.

I'd also suggest avoiding the need to match mixed-case: Perhaps `metric_buckettypes` and `metric_bucketsizes`? (Or ToLower them before lookup maybe)

Reply:
Changed to non-mixed case. I think to-lowering may be a bit challenging with struct tags. It would make parsing out fields where the user supplied non-lowercase values for metrics tags a bit wonky.